source/agora/consensus/EnrollmentManager.d

```diff
@@ -1264,28 +1266,29 @@ unittest
     assert(enrolls.length == 1);
 
     // make the enrollment a validator
-    man.addValidator(enroll, key_pair.address, Height(11), &utxo_set.peekUTXO, utxos);
+    man.addValidator(enroll, key_pair.address, Height(10), &utxo_set.peekUTXO, utxos);
```

Comment:
I don't see why the `Height` changed into 10. The part of the code means that the enrollments are nominated at the height of 10 and enrolled at the height of 11, and there are no enrollments in the pool after enrollment. So this should be 11. This change does not match with the basic flow of the enrollment process.

Reply:
This test actually enrolls the Validator one cycle late. I dont know if that was intentional. But with the new rules, this is not possible. All the height changes below are required because of this change, you cant enroll with the same enrollment more then once and you can use the enrollment only on a single height.

If a node does `createEnrollment(height = n)` and sends the request, that enrollment can only be included in the block #n. So the nodes create the enrollment for `Height(n)` at `Height(n-1)`.

See `checkAndEnroll`, it creates the enrollment for `block_height + 1` and that enrollment gets included in the next block.
https://github.com/bpfkorea/agora/blob/681fd853c1cb40c1df11959d47e0d81a9baca37d/source/agora/node/Validator.d#L466

Reply:
> If a node does `createEnrollment(height = n)` and sends the request, that enrollment can only be included in the block #n. So the nodes create the enrollment for `Height(n)` at `Height(n-1)`.
> 
> See `checkAndEnroll`, it creates the enrollment for `block_height + 1` and that enrollment gets included in the next block.

I see what you mean. I know the situation is kind of confusing because the `avail_height` remains. :) 
With your reasoning, the code below should be succeeded because you create an enrollment for Height(n) and try to enroll at the Height(n-1), so the calling 'getEnrollments' at the Height(n-1) should return 1.

`
 man.getEnrollments(enrolls, Height(9)); 
 assert(enrolls.length == 1);
`

Reply:
https://github.com/omerfirmak/agora/blob/681fd853c1cb40c1df11959d47e0d81a9baca37d/source/agora/consensus/EnrollmentManager.d#L248

`EnrollmentManager.getEnrollments` does the +1 trick as well.

https://github.com/omerfirmak/agora/blob/681fd853c1cb40c1df11959d47e0d81a9baca37d/source/agora/node/Ledger.d#L446

`Ledger` calls `getEnrollments` with the current height, `EnrollmenManager` request the enrollments valid for next block from the `EnrollmentPool`